airbyte-integrations/connectors/source-typeform/source_typeform/source.py

```diff
@@ -0,0 +1,230 @@
+
+    def path(
+        self,
+        stream_state: Mapping[str, Any] = None,
+        stream_slice: Mapping[str, Any] = None,
+        next_page_token: Mapping[str, Any] = None,
+    ) -> str:
+        return "/forms"
+
+    def next_page_token(self, response: requests.Response) -> Optional[int]:
+        page = self.get_current_page_token(response.url)
+        # stop pagination if current page equals to total pages
+        return None if not page or response.json()["page_count"] <= page else page + 1
+
+    def get_current_page_token(self, url: str) -> Optional[int]:
+        """
+        Fetches page query parameter from URL
+        """
+        parsed = urlparse.urlparse(url)
+        page = parse_qs(parsed.query).get("page")
+        return int(page[0]) if page else None
+
+    def request_params(
+        self,
+        stream_state: Mapping[str, Any],
+        stream_slice: Mapping[str, any] = None,
+        next_page_token: Mapping[str, Any] = None,
+    ) -> MutableMapping[str, Any]:
+        params = {"page_size": self.limit}
+        params["page"] = next_page_token or 1
+        return params
+
+
+class StreamMixin:
+    def stream_slices(self, **kwargs) -> Iterable[Optional[Mapping[str, any]]]:
+        for item in TrimForms(**self.config).read_records(sync_mode=SyncMode.full_refresh):
+            yield {"form_id": item["id"]}
+
+        yield from []
+
+
+class Forms(StreamMixin, TypeformStream):
+    """
+    This stream is responsible for detailed information about Form.
+    API doc: https://developer.typeform.com/create/reference/retrieve-form/
+    """
+
+    primary_key = "id"
+
+    def path(
+        self,
+        stream_state: Mapping[str, Any] = None,
+        stream_slice: Mapping[str, Any] = None,
+        next_page_token: Mapping[str, Any] = None,
+    ) -> str:
+        return f"/forms/{stream_slice['form_id']}"
+
+    def parse_response(self, response: requests.Response, **kwargs) -> Iterable[Mapping]:
+        yield response.json()
+
+
+class IncrementalTypeformStream(TypeformStream, ABC):
+    cursor_field: str = "submitted_at"
+    token_field: str = "token"
+
+    @property
+    def limit(self):
+        return super().limit
+
+    state_checkpoint_interval = limit
+
+    @abstractmethod
+    def get_updated_state(
+        self,
+        current_stream_state: MutableMapping[str, Any],
+        latest_record: Mapping[str, Any],
+    ) -> Mapping[str, Any]:
+        pass
+
+    def next_page_token(self, response: requests.Response) -> Optional[str]:
+        items = response.json()["items"]
+        if items and len(items) == self.limit:
+            return items[-1][self.token_field]
+        return None
+
+
+class Responses(StreamMixin, IncrementalTypeformStream):
+    """
+    This stream is responsible for fetching responses for particulat form_id.
+    API doc: https://developer.typeform.com/responses/reference/retrieve-responses/
+    """
+
+    primary_key = "response_id"
+    limit: int = 1000
```

Comment:
is it problematic to generate more than 1000 responses? to test API behaviur

Reply:
there no API to generate responses. And as I understand we have 250 responses per month limit https://github.com/airbytehq/airbyte/issues/3988#issuecomment-867600011

Reply:
Is there a reason we chose 1000?